app/assets/javascripts/shared/editor_toolbar.js

```diff
@@ -0,0 +1,232 @@
+            <i class="fa fa-quote-left"></i>\
+          </div>\
+          <div class="editor-btn" data-btn="link" aria-label="link">\
+            <i class="fa fa-link"></i>\
+          </div>\
+          <div class="editor-btn" data-btn="table" aria-label="table">\
+            <i class="fa fa-table"></i>\
+          </div>\
+          <div class="divider-vertical"></div>\
+          <div class="editor-btn" data-btn="block-code" aria-label="code block">\
+            <i class="fa fa-code"></i>\
+          </div>\
+          <div class="editor-btn" data-btn="inline-code" aria-label="inline code">\
+            <i class="fa fa-terminal"></i>\
+          </div>\
+          <div class="divider-vertical"></div>\
+          <div class="editor-btn" data-btn="list-ul" aria-label="unordered list">\
+            <i class="fa fa-list-ul"></i>\
+          </div>\
+          <div class="editor-btn" data-btn="list-ol" aria-label="ordered list">\
+            <i class="fa fa-list-ol"></i>\
+          </div>\
+        ');
+      }
+      else if ($(this).prev('input[type=text]').length) { // toolbar for input type=text elements
+        $(this).append(
+          '<div class="editor-btn" data-btn="bold" aria-label="bold text">\
+            <i class="fa fa-bold"></i>\
+          </div>\
+          <div class="editor-btn px-2" data-btn="italic" aria-label="italic text">\
+            <i class="fa fa-italic"></i>\
+          </div>\
+          <div class="editor-btn" data-btn="highlight" aria-label="highlighted text">\
+            <i class="fa fa-paint-brush"></i>\
+          </div>\
+          <div class="divider-vertical"></div>\
+          <div class="editor-btn" data-btn="link" aria-label="link">\
+            <i class="fa fa-link"></i>\
+          </div>\
+          <div class="editor-btn" data-btn="inline-code" aria-label="inline code">\
+            <i class="fa fa-terminal"></i>\
+          </div>\
+        ');
+      }
+    });
+
+    this.behaviors();
+  }
+
+  behaviors() {
+    var that = this;
+
+    // enabling/disabling specific toolbar functions for textareas on selection
+    $('[data-behavior~=editor-field] textarea').on('click change keyup select', function() {   
+      if (window.getSelection().toString().length > 0 || this.selectionStart != this.selectionEnd) { // when there is text selected
+        $(this).parent().find('[data-btn~=table]').addClass('disabled');
+      }
+      else { // when there is no text selected
+        $(this).parent().find('[data-btn~=table]').removeClass('disabled');
+      }
+    });
+
+    // when a toolbar button is clicked
+    $('[data-btn]').click(function () {
+      var $element = $(this).parents('[data-behavior~=editor-field]').children('textarea, input[type=text]');
+      var prefix = '', suffix = '', placeholder = '';
+  
+      // set the appropriate prefix, placeholder and suffix depending on which button was clicked
+      switch ($(this).data('btn')) {
+        case 'header':
```

Comment:
Same idea here. Big case statements make my brow sweat. 

Instead we could define a dictionary of element data. In this case it could be a plain javascript object with the appropriate naming.

```javascript
affixes = {}
affixes['bold'] = {'prefix': '*', placeholder: 'suffix', 'suffix': '*'}
// cont...
```

then instead of the case statement we can simply reach into the dictionary for the settings:
```javascript
affixes[$(this).data('btn')].prefix
> *
```